assets/js/googlesitekit/data/create-settings-store.test.js

```diff
@@ -0,0 +1,453 @@
+	} );
+
+	afterAll( () => {
+		API.setUsingCache( true );
+	} );
+
+	afterEach( () => {
+		unsubscribeFromAll( registry );
+		apiFetchSpy.mockRestore();
+	} );
+
+	describe( 'name', () => {
+		it( 'returns the correct default store name', () => {
+			expect( storeDefinition.STORE_NAME ).toEqual( `${ STORE_ARGS[ 0 ] }/${ STORE_ARGS[ 1 ] }` );
+		} );
+	} );
+
+	describe( 'actions', () => {
+		describe( 'setSettings', () => {
+			it( 'requires the values param', () => {
+				expect( () => {
+					dispatch.setSettings();
+				} ).toThrow( 'values is required.' );
+			} );
+
+			it( 'updates the respective settings', () => {
+				const values1 = { setting1: 'old', setting2: 'old' };
+				const values2 = { setting2: 'new' };
+
+				dispatch.setSettings( values1 );
+				expect( store.getState().settings ).toMatchObject( values1 );
+
+				dispatch.setSettings( values2 );
+				expect( store.getState().settings ).toMatchObject( { ...values1, ...values2 } );
+			} );
+		} );
+
+		describe( 'fetchSettings', () => {
+			it( 'does not require any params', () => {
+				expect( () => {
+					dispatch.fetchSettings();
+				} ).not.toThrow();
+			} );
+		} );
+
+		describe( 'receiveSettings', () => {
+			it( 'requires the values param', () => {
+				expect( () => {
+					dispatch.receiveSettings();
+				} ).toThrow( 'values is required.' );
+			} );
+
+			it( 'receives and sets values', () => {
+				const serverValues = { setting1: 'serverside', setting2: 'serverside' };
+				const clientValues = { setting1: 'clientside' };
+
+				dispatch.setSettings( clientValues );
+				dispatch.receiveSettings( serverValues );
+
+				// Client values take precedence if they were already modified before receiving from the server.
+				expect( store.getState().settings ).toMatchObject( { ...serverValues, ...clientValues } );
+			} );
+		} );
+
+		describe( 'saveSettings', () => {
+			it( 'does not require any params', () => {
+				expect( async () => {
+					fetch
+						.doMockOnceIf(
+							/^\/google-site-kit\/v1\/core\/site\/data\/settings/
+						)
+						.mockResponseOnce(
+							JSON.stringify( { setting1: 'serverside' } ),
+							{ status: 200 }
+						);
+
+					await dispatch.saveSettings();
+				} ).not.toThrow();
+			} );
+
+			it( 'updates settings from server', async () => {
+				const response = { setting1: 'serverside' };
+				fetch
+					.doMockIf(
+						/^\/google-site-kit\/v1\/core\/site\/data\/settings/
+					)
+					.mockResponse(
+						JSON.stringify( response ),
+						{ status: 200 }
+					);
+
+				// The server is the authority. So because this won't be part of the response
+				// (see above), it will be disregarded.
+				dispatch.setSettings( { setting2: 'clientside' } );
+
+				await dispatch.saveSettings();
+				// Two fetch requests, one to get settings, the other to update.
+				expect( fetch ).toHaveBeenCalledTimes( 2 );
+
+				expect( store.getState().settings ).toMatchObject( response );
```

Comment:
Interesting. I consider this unexpected behaviour. The client (should) fetch settings before setting them, but what if you present the user an interface where the can change parts of the data on the client without server involvement? I suppose I'm struggling to find an example other than a list where they could add items while existing ones were fetched, but it's a bit surprising this happens.

That said: it's codified and thus expected, so that's good. And I suppose we likely want loading screens that prevent user interaction until they can save settings, so fair enough 👍